src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilter.java

```diff
@@ -0,0 +1,90 @@
+package com.puppycrawl.tools.checkstyle.filters;
+
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import com.puppycrawl.tools.checkstyle.api.*;
+
+public class SuppressionXpathSingleFilter extends AutomaticBean implements
+        Filter, ExternalResourceHolder {
+
+    /** The regexp to match check names against. */
+    private Pattern checkRegexp;
+
+    /** The pattern for check class names. */
+    private String checkPattern;
+
+    /** The regexp to match message names against. */
+    private Pattern messageRegexp;
+
+    /** The pattern for message names. */
+    private String messagePattern;
+
+    /** Module id filter. */
+    private String moduleId;
+
+    /** Xpath query. */
+    private String xpathQuery;
+
+    /** Set of individual suppresses. */
+    private FilterSet filters = new FilterSet();
```

Comment:
I don't believe this should have any FilterSets. It should be a self-contained filter.